### ISSUE TITLE
System.IO.FileNotFoundException: Could not load file or assembly (windows)

### ISSUE THREAD

**Issue by jalbarl** [mono/mono#6301]
## Steps to Reproduce

1. Download the Intel-SA-00086 Detection Tool (https://downloadcenter.intel.com/download/27150?v=t)
2. Extract the zip file and run the console application from the DiscoveryTool folder: mono.exe Intel-SA-00086-console.exe -n -f -c 

## Current Behavior
The application won't start, because mono cannot find a file (it might be embedded into the binary - because it is not present on the filesystem)

## Expected Behavior
The application is working fine on a separate machine, which has .Net 4.5 installed. I may be trying to do something that is unsupported, or I just ran into the "not implemented" area. I tried to find more information about this kind of error message, but it is still unclear if the application should start correctly or not (maybe adding some more information on the stack trace would help debug the issue).

### On which platforms did you notice this

[ ] macOS
[ ] Linux
[ X] Windows (7 x64)

**Version Used**:

Mono JIT compiler version 5.4.1 (Visual Studio built mono)
This is the standard x64 version downloaded from http://www.mono-project.com
I initially used version 5.0.0, with the same result.

### Stacktrace

```
Unhandled Exception:
System.IO.FileNotFoundException: Could not load file or assembly 'Intel-SA-00086-console-DLL, Version=1.0.0.146, Culture=neutral, PublicKeyToken=null' or one of its dependencies.
File name: 'Intel-SA-00086-console-DLL, Version=1.0.0.146, Culture=neutral, PublicKeyToken=null'
[ERROR] FATAL UNHANDLED EXCEPTION: System.IO.FileNotFoundException: Could not load file or assembly 'Intel-SA-00086-console-DLL, Version=1.0.0.146, Culture=neutral, PublicKeyToken=null' or one of its dependencies.
File name: 'Intel-SA-00086-console-DLL, Version=1.0.0.146, Culture=neutral, PublicKeyToken=null'
```

**Comment by lambdageek** [comment]
`Intel-SA-00086-console-DLL` is embedded in `Intel-SA-00086-console.exe` as a manifest resource.  `ikdasm Intel-SA-00086-console.exe` shows:

```
.mresource public 'DiscoveryTool.CLI.Intel-SA-00086-console-DLL.dll'
{
  // Offset: 0x00418630 Length: 0x00016000
  // WARNING: managed resource file 'DiscoveryTool.CLI.Intel-SA-00086-console-DL
L.dll' created
}
```

I'm not sure whether Mono's loader can load assemblies from manifest resources.

**Comment by migueldeicaza** [comment]
I think that the problem is that we are trying to resolve types at startup before running some static constructors.

The Main type has a static constructor that calls some extraction code (`DiscoveryTool.FileExtractor`), but this code does not get a chance to run, it is supposed to be triggered by the static constructor of the `Program` class that initializes it.

Something in Program is triggering the assembly resolution to take place before this static constructor runs.

**Comment by jalbarl** [comment]
You are right. I found a tool that is able to extract such resources (ILSpy), and placed them next to the executable. Not it is loaded successfully, but only until the next error (see below).

Should I open a new Issue for it, or this also a known limitation?
I think that the code is trying to lookup "Administrators", but the expected name is "BUILTIN\Administrators" (from https://github.com/mono/mono/blob/master/mcs/class/corlib/System.Security.Principal/WellKnownAccount.cs). Maybe WellKnownAccount.LookupByName could be improved to cater for such naming variations.

Unhandled Exception:
System.Security.Principal.IdentityNotMappedException: Cannot map account name: Administrators
  at System.Security.Principal.NTAccount.Translate (System.Type targetType) [0x0004d] in <ee9b80546f454fd48a31a390370231ef>:0
  at System.Security.AccessControl.ObjectSecurity.SidFromIR (System.Security.Principal.IdentityReference identity) [0x00014] in <ee9b80546f454fd48a31a390370231ef>:0
  at System.Security.AccessControl.CommonObjectSecurity.ModifyAccess (System.Security.AccessControl.AccessControlModification modification, System.Security.AccessControl.AccessRule rule, System.Boolean& modified) [0x00051] in <ee9b80546f454fd48a31a390370231ef>:0
  at System.Security.AccessControl.CommonObjectSecurity.AddAccessRule (System.Security.AccessControl.AccessRule rule) [0x00000] in <ee9b80546f454fd48a31a390370231ef>:0
  at System.Security.AccessControl.FileSystemSecurity.AddAccessRule (System.Security.AccessControl.FileSystemAccessRule rule) [0x00000] in <ee9b80546f454fd48a31a390370231ef>:0
  at DiscoveryTool.FileExtractor.Extract (System.String directoryName, System.Boolean localization) [0x0001f] in <e9d0a33a973d4799a512096db43e7db3>:0
  at DiscoveryTool.CLI.Program.Main (System.String[] args) [0x000f6] in <e9d0a33a973d4799a512096db43e7db3>:0

**Comment by lambdageek** [comment]
Thanks, @migueldeicaza, you're right, looks like the static constructor doesn't run early enough.
Looks like `Main` has a local variable with type: `['Intel-SA-00086-console-DLL']DiscoveryTool.logWriter V_3` - so the referenced assembly load is triggered when we try to JIT `Main`.

@jalbarl Please make a separate issue for the other exception.

**Comment by lambdageek** [comment]
Minimal example to reproduce:

File `BBB.cs`:
```csharp
public class BBB {
	public BBB () {
		System.Console.WriteLine ("BBB");
	}
}
```

File `SMain.cs`:
```csharp
using System;


public class Foo {
	static Foo () {
		System.IO.File.Copy ("AAA.dll", "BBB.dll", true);
	}

	public static void Main () {
		var x = new BBB ();
		Console.WriteLine ("Main!");
	}
}
```

To build and run:
```
csc -target:library BBB.cs
csc -r:BBB.dll SMain.cs

mv BBB.dll AAA.dll
mono SMain.exe
```

Expected behavior:  Should print:
```
BBB
Main!
```

Actual behavior:
```
Unhandled Exception:
System.IO.FileNotFoundException: Could not load file or assembly 'BBB, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies.
File name: 'BBB, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'
[ERROR] FATAL UNHANDLED EXCEPTION: System.IO.FileNotFoundException: Could not load file or assembly 'BBB, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies.
File name: 'BBB, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'
```
